pkg/release/release.go

```diff
@@ -4,14 +4,26 @@ import (
 	"context"
 	"fmt"
 	"strings"
+	"time"
 
 	releasev1alpha1 "github.com/giantswarm/apiextensions/v2/pkg/apis/release/v1alpha1"
 	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	gocache "github.com/patrickmn/go-cache"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func GetComponentVersionsFromRelease(ctx context.Context, ctrlClient client.Client, releaseVersion string) (map[string]string, error) {
+var (
+	cacheExpiry = 24 * time.Hour
+
+	// Cache of release components
+	// key: release version
+	// value: map "component name": "component version"
+	releaseComponentsCache = gocache.New(cacheExpiry, 24*time.Hour)
```

Comment:
Is there a mechanism that refreshes the cached release if it changes? I'm afraid this caching is a dangerous footgun during development.

Could there be a simple watcher running in the background instead, that would persist & refresh present `Releases` in an in-memory map, which would then be served like the API below?

Reply:
> Is there a mechanism that refreshes the cached release if it changes?

AFAIK there isn't (I did not implement anything in azure-admission-controller and I don't think there is anything like that in the cache library).

> I'm afraid this caching is a dangerous footgun during development.
I can't say I understand why :confused:

I believe that when the cache item expires the `cache.Get(key)` will not return any result. When that happens `azure-admission-controller` will just call API server again and cache the results again for the next 24h, here:

https://github.com/giantswarm/azure-admission-controller/pull/247/files/b53c0880f1affe47aaf7d45950332baed6d85302#diff-c1b95cb221c72b791b38018615bb6bb835b5f303ff0cd75871cbb1f8edbb8514R67-R81

Maybe there is a better cache expiry time than 24h, this was just a semi-wild semi-educated guess :)

Reply:
Sorry, I should have been more elaborate:

When the release is cached for some period of time, it affects the behavior and defaulting of admission controller here. For most of the time it works fine and we forget about it. Then at some point we need to do some bisecting or frequent updating of component for a given release and we observe weird behavior. If we don't remember that there exists this cache inside admission controller (and that it requires re-start of the whole program in order to flush it), we can easily waste a lot of time looking elsewhere for the root cause of this weird behavior.

If the implementation had watch + map instead of on-demand cache, the releases would be effectively cached, but also always up-to-date.

I'm not strictly requiring this, but just saying that there is a potential here for overlooking once we forget the existence of this cache :slightly_smiling_face:

Reply:
Oh I see now, so while we are working on a component `X`, frequently changing its version in a release during develpment and its version in the release is cached in the azure-admission-controller, weird stuff might happen because that version will not be refreshed in the cache. Makes sense.

What kind of watch implementation would you suggest?

I just wouldn't like to complicate it to the point where it's just better not to do caching at all, because its benefits would be almost lost due to increased complexity of the implementation which adds another type of burden and issues :smile: OTOH simple but flaky behavior (like you described) is also not desireable.